layout/src/adt/dag.rs

```diff
@@ -459,14 +794,647 @@ impl DAG {
             }
         }
 
-        // For each node in the order.
+        levels
+    }
+
+    pub(crate) fn place_horizontal_borders(
+        &mut self,
+        subgraph_levels: &Vec<(usize, usize)>,
+    ) {
+        let num_levels = self.num_levels();
+        self.top_border_ranks = vec![vec![]; num_levels];
+        self.bottom_border_ranks = vec![vec![]; num_levels];
+
+        self.top_border_ranks[0] = vec![vec![SubgraphHandle::new(0)]];
+        self.bottom_border_ranks[num_levels - 1] =
+            vec![vec![SubgraphHandle::new(0)]];
+
+        let mut inner_levels = vec![None; self.num_subgraphs()];
+
+        inner_levels[0] = Some((0, 0));
+        for s in 0..self.num_subgraphs() {
+            self.new_horizontal_border(
+                SubgraphHandle::new(s),
+                subgraph_levels,
+                &mut inner_levels,
+            );
+        }
+    }
+
+    fn new_horizontal_border(
+        &mut self,
+        subgraph_idx: SubgraphHandle,
+        levels_map: &Vec<(usize, usize)>,
+        inner_levels: &mut Vec<Option<(usize, usize)>>,
+    ) {
+        let (p_lvl_s, p_lvl_e) = levels_map[subgraph_idx.idx];
+
+        let (p_in_lvl_s, p_in_lvl_e) = inner_levels[subgraph_idx.idx].expect(
+            "Ordering subgraph horizontal borders is wrong, this is a bug.",
+        );
+
+        let mut c_in_lvl_s;
+        let mut c_in_lvl_e;
+
+        for child_s in self.nesting_tree.subgraphs[subgraph_idx.idx]
+            .subgraphs
+            .iter()
+        {
+            let (lvl_s, lvl_e) = levels_map[child_s.idx];
+            if lvl_s == p_lvl_s {
+                if p_in_lvl_s + 1 < self.top_border_ranks[lvl_s].len() {
+                    self.top_border_ranks[lvl_s][p_in_lvl_s + 1].push(*child_s);
+                } else {
+                    self.top_border_ranks[lvl_s].push(vec![*child_s]);
+                }
+
+                c_in_lvl_s = p_in_lvl_s + 1;
+            } else {
+                if self.top_border_ranks[lvl_s].is_empty() {
```

Comment:
1. Here's where the panic for `lvl_s` being out of bounds for `top_border_ranks` happens when there's an empty graph.